core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js

```diff
@@ -419,6 +421,11 @@ oppia.directive('conversationSkin', [
             AutogeneratedAudioPlayerService.cancel();
             var index = PlayerPositionService.getActiveCardIndex();
             $scope.activeCard = PlayerTranscriptService.getCard(index);
+            if ($scope.activeCard.conceptCard) {
```

Comment:
I am concerned about tacking on "conceptCard" to activeCard like this. We should never just modify a dict on the fly like this. It's hard to tell what's in the dict.

Perhaps make activeCard into a domain object if you need to modify it further. It seems like it's getting sufficiently complex to warrant that.

Also, something might be wrong with the naming. It seems weird for a card to be a child of a card. Isn't the active card the concept card itself?

Finally, I feel like you want a bit more structure around this. Conceptually, there are at least two ways to think about this: 
- You have one type of card. But either its content comes from the exploration data or it is generated on the fly (in this case, from the skill).
- You have two types of cards: concept cards and exploration cards.

If you go with the former then make "activeCard" into a domain object (probably just Card or maybe CardContents) and populate it accordingly, at the right times. It doesn't need to know whether it's a concept card or not.

If you go with the latter then you need two types of objects to represent cards: ConceptCard and ExplorationCard. Then you'd create those objects (with their own properties).

I think I lean towards the former option (less duplication) but this is something you'll want to decide. If you end up having separate UIs altogether for both types of cards, then maybe treating them as two different types of objects would be better. But if they're all just being represented as "card" then a single domain object should suffice and this also has the virtue of being simpler.

Reply:
I had initially thought of adding explanation in contenthtml (a field in activeCard), and adding a workedExamples field with all other fields null in activeCard, but the issue was that there are some functions in ConversationSkinDirective that are continuously called by directives in conversation_skin_directive.html that depend on exploration specific properties. So, changing any existing fields in activeCard led to a lot of errors, that's why I added a conceptCard field to make it separate and not interfere with exploration properties. 
Actually, reading the last comment, this problem wouldn't arise if concept card directive was called outside conversation skin. How about I add <concept-card> to exploration_player.html and create a controller for the html (it doesn't have one now) that then renders either <conversation-skin> or <concept-card> depending on answer submitted. I think this way, normal cards and concept card can be modelled to be different entities and not have one inside another. Would that work? since, other than the skill id linked to submitted answer group (which could be obtained through a $broadcast or other methods), concept card is independent of anything else in the exploration.

Reply:
But I think if we are now modelling concept cards as basically being like exploration cards but with different content (and the worked examples part of it as separate cards that come after) then you'll want a single object for Card that can represent both, right? So I feel like you want to do something more like the one-object approach where you populate a generic card object with data, rather than treat the concept card as a separate concept altogether -- it doesn't seem like we're planning to treat it that way from the user perspective.

Btw, one other thought re staging this. I think in the initial phase you can just do something that shows only the "review" part of the concept card and doesn't include the worked examples. We can add the latter in a separate pass once we have @christineycc's mocks. I think that would be easier to handle since introducing a new type of card in the exploration player is quite a big conceptual change.

Reply:
Ok, yeah, then shall I remove the concept card directive, and show the review material as a state's content html, but having a continue button interaction?
But here also if I try to merge it with ConversationSkinDirective, since concept card doesn't have a state name, some functions will throw an error. I think this can be resolved in 2 ways.
1. Specify a state name (like 'conceptCard') for these and have a check in all functions to return if this is the state name.
2. Have the concept card have the same state name as the card from which it was called.  

Second method doesn't require all the checks in the various functions so, it might be better.
Shall I follow this?

Reply:
Since the card that shows the contents (tutor-card) is called from conversation-skin, these checks would still be a problem. In that case, maybe I can have stateName as null, and add a check to all those functions to not proceed further if stateName is null? That would work also.
I can reuse tutor-card to show concept card contents. These checks for stateName are the only issue, so far.

Reply:
I think a small refactor might be needed here. Why does tutor card need the state name? Especially if it's just displaying the data in the card?

If it's for submitting answers, then perhaps the tutor card should just submitAnswer to its conversation skin directive, which attaches the state name and sends it on (or does other things depending on whether the current card is a concpet card or not).

In general, it's also worth thinking about how you'd design the components here if you were building them from scratch, with all the information you already have. Don't take the existing design and try to tack stuff on -- you'll end up with an unmaintainable mess. Figure out what the ideal solution is (conceptually) and then we can figure out how to move incrementally towards it.

Reply:
Oh, sorry, I think I wasn't clear. Tutor card doesn't need state name, but it is called from conversation skin which does. So, even if I modify tutor card to show concept card contents, because it is called from conversation skin, the state name would still be an issue.
Yeah, tutor card does do that currently. Submit answer is indeed handled by conversation skin.

Reply:
Ah OK. Well, conversation skin should be able to handle that then, in the sense that it should be able to track whether we're currently in a "normal state playing" context or a "concept card" context, and only use the state name in the former.

My suggestions in the last paragraph still apply -- think about the ideal way to write the conversation skin directive, if you had known about these two concepts at the outset. You could perhaps move some of the functionality to PlayerServices.js if that helps to simplify things.

Reply:
Done,  removed the concept card directive and modified tutor card and conversation skin directive to show concept card contents (and hide no interaction message when doing so). 
Also, removed the extra conceptCard field from transcript and added the skill explanation as a state contentHtml and made stateName as null when conceptCard is being displayed.